Devices/Adafruit MMA8451 - Accelerometer/Adafruit_MMA8451_SS/Adafruit_MMA8451.cpp

```diff
@@ -0,0 +1,301 @@
+  i2cwrite((uint8_t)reg);
+  i2cwrite((uint8_t)(value));
+  Wire.endTransmission();
+}
+
+/**************************************************************************/
+/*!
+    @brief  Public function to write 8-bits to the specified destination register
+*/
+/**************************************************************************/
+void Adafruit_MMA8451::writeRegister8_public(uint8_t reg, uint8_t value) {
+  uint8_t ctl1 = readRegister8(MMA8451_REG_CTRL_REG1);
+  writeRegister8(MMA8451_REG_CTRL_REG1, 0x00);            // deactivate
+  writeRegister8(reg, value);
+  writeRegister8(MMA8451_REG_CTRL_REG1, ctl1 | 0x01);     // activate
+  
+}
+
+/**************************************************************************/
+/*!
+    @brief  Reads 8-bits from the specified register
+*/
+/**************************************************************************/
+uint8_t Adafruit_MMA8451::readRegister8(uint8_t reg) {
+    
+//undocumented version of requestFrom handles repeated starts on Arduino Due
+#ifdef __SAM3X8E__
+    Wire.requestFrom(_i2caddr, 1, reg, 1, true);
+#else
+    //I don't know - maybe the other verion of requestFrom works on all platforms.
+    //  honestly, I don't want to go through and test them all.  Doing it this way
+    //  is already known to work on everything else
+    Wire.beginTransmission(_i2caddr);
+    i2cwrite(reg);
+    Wire.endTransmission(false); // MMA8451 + friends uses repeated start!!
+    Wire.requestFrom(_i2caddr, 1);
+#endif
+    
+    if (! Wire.available()) return -1;
+    return (i2cread());
+}
+
+/**************************************************************************/
+/*!
+    @brief  Instantiates a new MMA8451 class in I2C mode
+*/
+/**************************************************************************/
+Adafruit_MMA8451::Adafruit_MMA8451(int32_t sensorID) {
+  _sensorID = sensorID;
+}
+
+/**************************************************************************/
+/*!
+    @brief  Setups the HW (reads coefficients values, etc.)
+*/
+/**************************************************************************/
+bool Adafruit_MMA8451::begin(uint8_t i2caddr) {
+  Wire.begin();
+  _i2caddr = i2caddr;
+
+  /* Check connection */
+  uint8_t deviceid = readRegister8(MMA8451_REG_WHOAMI);
+  if (deviceid != 0x1A)
+  {
+    /* No MMA8451 detected ... return false */
+    //Serial.println(deviceid, HEX);
+    return false;
+  }
+
+  writeRegister8(MMA8451_REG_CTRL_REG2, 0x40); // reset
+
+  while (readRegister8(MMA8451_REG_CTRL_REG2) & 0x40);
+
+  // enable 4G range
+  writeRegister8(MMA8451_REG_XYZ_DATA_CFG, MMA8451_RANGE_4_G);
+  // High res
+  writeRegister8(MMA8451_REG_CTRL_REG2, 0x02);
+  // DRDY on INT1
+  writeRegister8(MMA8451_REG_CTRL_REG4, 0x01);
+  writeRegister8(MMA8451_REG_CTRL_REG5, 0x01);
+
+  // Turn on orientation config
+  writeRegister8(MMA8451_REG_PL_CFG, 0x40);
+
+  // Activate at max rate, low noise mode
+  writeRegister8(MMA8451_REG_CTRL_REG1, 0x01 | 0x04);
```

Comment:
Does 0x1 | 0x4 select both the 0th and 2nd bit from the right, and is this different than assigning 0x5?

Reply:
@kwonma there is no difference, it just stylistic so that when someone else is reading the code it is explicitly setting two bits in a register as opposed to one value.